pkg/controller/workloadspread/workloadspread_controller.go

```diff
@@ -318,23 +322,81 @@ func getInjectWorkloadSpreadFromPod(pod *corev1.Pod) *wsutil.InjectWorkloadSprea
-		if injectWS == nil || injectWS.Name != ws.Name {
-			continue
+		if subsetName, err := r.getSuitableSubsetNameForPod(ws, pods[i]); err != nil {
+			return nil, err
+		} else if _, exist := podMap[subsetName]; exist {
+			podMap[subsetName] = append(podMap[subsetName], pods[i])
+		}
+	}
+
+	return podMap, nil
+}
+
+// getSuitableSubsetNameForPod will return ("", nil) if not found suitable subset for pod
+func (r *ReconcileWorkloadSpread) getSuitableSubsetNameForPod(ws *appsv1alpha1.WorkloadSpread, pod *corev1.Pod) (string, error) {
+	injectWS := getInjectWorkloadSpreadFromPod(pod)
+	if injectWS == nil || injectWS.Name != ws.Name {
+		// process the pods that were created before workloadSpread
+		matchedSubset, err := r.getSuitableSubsetForOldPod(ws, pod)
+		if err != nil {
+			return "", err
+		} else if matchedSubset == nil {
+			return "", nil
+		}
+		return matchedSubset.Name, nil
+	}
+	return injectWS.Subset, nil
+}
+
+// getSuitableSubsetForOldPod returns an suitable subset for the pod which was created before workloadSpread.
+// getSuitableSubsetForOldPod will return (nil, nil) if there is no suitable subset for the pod.
+func (r *ReconcileWorkloadSpread) getSuitableSubsetForOldPod(ws *appsv1alpha1.WorkloadSpread, pod *corev1.Pod) (*appsv1alpha1.WorkloadSpreadSubset, error) {
+	if len(pod.Spec.NodeName) == 0 {
+		return nil, nil
+	}
+
+	node := &corev1.Node{}
+	if err := r.Get(context.TODO(), types.NamespacedName{Name: pod.Spec.NodeName}, node); err != nil {
+		if errors.IsNotFound(err) {
+			return nil, nil
 		}
-		if _, exist := podMap[injectWS.Subset]; exist {
-			podMap[injectWS.Subset] = append(podMap[injectWS.Subset], pods[i])
+		return nil, err
+	}
+	for i := range ws.Spec.Subsets {
+		subset := &ws.Spec.Subsets[i]
+		// in case of that this pod was scheduled to the node which matches a subset of workloadSpread
+		// TODO: check the patch of subset, one can have subset that only have patch but without any affinity or toleration changes.
+		if matched, err := nodeMatchesSubset(node, subset); err != nil {
+			// requiredSelectorTerm field was validated at webhook stage, so this error should not occur
+			// this error should not be returned, because it is a non-transient error
+			klog.Errorf("unexpected error occurred when matching node with subset, please check requiredSelectorTerm field of subset (%s) in WorkloadSpread (%s/%s), err: %s",
+				subset.Name, ws.Namespace, ws.Name, err.Error())
+		} else if matched {
+			// add MatchedWorkloadSpreadSubsetAnnotations for this pod based on the matched subset
+			// the purpose of this operation is to avoid to do this match again in the next reconcile
+			by, _ := json.Marshal(&wsutil.InjectWorkloadSpread{Name: ws.Name, Subset: subset.Name})
+			if pod.Annotations == nil {
+				pod.Annotations = make(map[string]string)
+			}
+			pod.Annotations[wsutil.MatchedWorkloadSpreadSubsetAnnotations] = string(by)
+			if err := r.Update(context.TODO(), pod); err != nil {
```

Comment:
You'd better use `patch` instead of `update` to avoid conflict.

Reply:
done